apps/emqx_dashboard/src/emqx_dashboard_monitor.erl

```diff
@@ -207,19 +209,79 @@ code_change(_OldVsn, State = #state{}, _Extra) ->
 %% -------------------------------------------------------------------------------------------------
 %% Internal functions
 
+all_data() ->
+    Fn = fun(#emqx_monit{time = Time, data = Data}, Acc) -> [{Time, Data} | Acc] end,
+    lists:keysort(1, ets:foldl(Fn, [], ?TAB)).
+
+inplace_downsample() ->
+    All = all_data(),
+    Now = erlang:system_time(millisecond),
+    Compacted = compact(Now, All, []),
+    {Deletes, Writes} = compare(All, Compacted, [], []),
+    {atomic, ok} = mria:transaction(
```

Comment:
I guess there's now a tiny chance observing inconsistencies if this coincides with `do_sample_local/1` (because the latter goes into the ETS directly). I think this could be minimized by applying changes in the timestamp order (e.g. by making `compare/3` produce a single `[{_TS :: timestamp(), _Data | delete}]`). Or avoided by forcing `do_sample_local/1` to go through gen_server, but that's probably an overkill.

Reply:
agree that it's overkill.
the downsample happens only every 10m,
ets:select may observe less data points during this transaction, but the next refresh will likely get a consistent view again.